lib/cucumber/core/event_bus.rb

```diff
@@ -0,0 +1,61 @@
+require 'cucumber/core/events'
+
+module Cucumber
+  module Core
+
+    # Event Bus
+    #
+    # Implements an in-process pub-sub event broadcaster allowing multiple observers
+    # to subscribe to events that fire as your tests are executed.
+    #
+    class EventBus
+      attr_reader :event_types
+
+      # @param [Hash{Symbol => Class}] a hash of event types to use on the bus
+      def initialize(registry = Events.registry)
+        @event_types = registry.freeze
+        @handlers = {}
+      end
+
+      # Register for an event. The handler proc will be called back with each of the attributes
+      # of the event.
+      def on(event_id, handler_object = nil, &handler_proc)
+        handler = handler_proc || handler_object
+        raise ArgumentError, "Please pass either an object or a handler block" unless handler
+        raise ArgumentError, "Please use a symbol for the event_id" unless event_id.is_a?(Symbol)
+        raise ArgumentError, "Event ID #{event_id} is not recognised. Try one of these:\n#{event_types.keys.join("\n")}" unless is_registered_id?(event_id)
+        event_class = event_types[event_id]
+        handlers_for(event_class) << handler
+      end
+
+      # Broadcast an event
+      def broadcast(event)
+        raise ArgumentError, "Event type #{event.class} is not registered. Try one of these:\n#{event_types.values.join("\n")}" unless is_registered_type?(event.class)
+        handlers = handlers_for(event.class)
+        handlers.each { |handler| handler.call(event) }
```

Comment:
Good point - this code was much longer at a point in the past, and that `handlers` temp is a hangover from that. Feel free to JFDI that change on master @danascheider 👍